### ISSUE TITLE
Legacy vegetation job & job-writing support

### ISSUE THREAD

**Issue by aleksa-krolls** [OpenFn/ConSoSci#117]
## Request
WCS needs to map [this legacy plant data](https://www.openfn.org/projects/p5x4g4/messages/mgk5v7dz) to their DB. @dianedetoeuf drafted an initial job to map this data, but needs some support to finish the job and ensure it runs successfully. We'll schedule time to walk through needed job changes with Diane next week, but in the meantime can you please...
1. Draft a second version of this job that _will_ work with this state. On the call we can then compare your job with Diane's to walk through what additional changes need to be made. 
2. Check out [this error](https://www.openfn.org/projects/p5x4g4/runs/rx85x877) encountered when she trying to test this... this is what I flagged on Slack. I expect the job to fail, but this seems like a strange error message... but maybe because the DB doesn't like that we're sending undefined objects? 

## state.json
See LP - `WCS Programs Database`
https://www.openfn.org/projects/p5x4g4/messages/mgk5v7dz

## adaptor
mssql

## expression.js
Feel free to run on prod, or see LP and use OpenFn VPN
https://www.openfn.org/projects/p5x4g4/jobs/jywrew
https://github.com/OpenFn/ConSoSci/blob/master/vegetation/legacyVegetation_vDD.js

**Comment by aleksa-krolls** [comment]
@lakhassane Ignore my #2! Figured it out...
@dianedetoeuf fyi, I realized that you selected the `postgres` adaptor for your job, rather than the `mssql` adaptor... this is why we were seeing the strange error message. 
Here is a more helpful error message... looks like the DB column name should be `Answer_ID`: 
https://www.openfn.org/projects/p5x4g4/runs/rg5ej5d6

**Comment by lakhassane** [comment]
@aleksa-krolls here is the draft of another version that "could" work https://github.com/OpenFn/ConSoSci/blob/master/vegetation/legacyVegetation_vDD_v2.js

The big take away is that data in the payload are inside arrays (general, liana_old). So you would need an `each`. 
I made some comments on the job to pay attention to as when we might need to use `findValue`.

**Comment by lakhassane** [comment]
hey @dianedetoeuf 
So basically what you sort of missed are the closing bracket and closing parenthesis of your `findValue`functions.
Instead of having this
```
WCSPROGRAMS_VegetationForestTypeID: findValue({
      uuid: 'WCSPROGRAMS_VegetationForestTypeID',
      relation: 'WCSPROGRAMS_VegetationForestType',
      where: {
        WCSPROGRAMS_VegetationForestTypeExtCode: state.handleValue(
          dataValue('Vegetation_field')(state) 
        ),
      },
```

it should says
```
WCSPROGRAMS_VegetationForestTypeID: findValue({
      uuid: 'WCSPROGRAMS_VegetationForestTypeID',
      relation: 'WCSPROGRAMS_VegetationForestType',
      where: {
        WCSPROGRAMS_VegetationForestTypeExtCode: state.handleValue(
          dataValue('Vegetation_field')(state) 
        ),
      },
}) // new line added
```

**Comment by lakhassane** [comment]
Now working with `findValue` can be tricky. If you run this and find any unconsistency on the DB please let me know because you might need to switch that operation that is using those findValue to the new way of writing it

**Comment by aleksa-krolls** [comment]
@dianedetoeuf As discussed, see the [Vegetation job](https://github.com/OpenFn/ConSoSci/blob/master/vegetation/VegetationClassficationAndTreeMeasurementForm.js#L85-L358) for inspiration for finishing the mappings where you will have something like...
```
alterState(async state => {  
  const vegMapping = {
    //your field mappings here
  };
  return upsert('WCSPROGRAMS_Vegetation', 'Answer_ID', vegMapping)(state);
});
```
Let me and @lakhassane know if you run into any further issues or questions.

**Comment by dianedetoeuf** [comment]
It looks like I've managed to use the alterstate(async ...) function but have an issue on the handleValue that I don't know how to solve :)

See [here](https://www.openfn.org/projects/p5x4g4/jobs/jywrew/edit?runId=r89b9kde) line 127.

**Comment by lakhassane** [comment]
Hey @dianedetoeuf sorry for the delay.

`state.handleValue` is a function so it should be created somewhere. If you look at the `vegetationClassification` job you can create something like this.
https://github.com/OpenFn/ConSoSci/blob/master/vegetation/VegetationClassficationAndTreeMeasurementForm.js#L45-L71

I drafted it on your job so you can see.

**Comment by lakhassane** [comment]
I also commented it out from line 15 to 125 and drafted a way for using `async/await `with `findValue` (in replacement of the operation below the one I drafted. It's a bit tricky because of mssql constraint. Feel free to replicate for that big operation that is using findValue again
https://github.com/OpenFn/ConSoSci/blob/master/vegetation/legacyVegetation_vDD.js#L127-L152

Beside that I run the job and it's failing because of invalid database column name: https://www.openfn.org/projects/p5x4g4/runs/rk53kany

**Comment by aleksa-krolls** [comment]
thanks for the feedback @lakhassane! 

@dianedetoeuf Fyi, I think most of the AnswerId columns are labeled as `Answer_ID` for these Veg tables. 
See the main Vegetation job for reference: https://github.com/OpenFn/ConSoSci/blob/master/vegetation/VegetationClassficationAndTreeMeasurementForm.js

**Comment by dianedetoeuf** [comment]
Hi, ok thanks for this new alternative! and fixed the Answer_ID. Seems like we have a new type of error: 
https://www.openfn.org/projects/p5x4g4/runs/rk53ky37

**Comment by lakhassane** [comment]
@dianedetoeuf I check the error and corrected. You were missing a `(state)`after the last parenthesis of findValue.

I run it locally but it was failing due to 2 issues now:
- `RequestError: Invalid column name 'PlotGPS'.` (I comment it out at line `27`)
- `Cannot insert the value NULL into column 'WCSPROGRAMS_VegetationName'` (it's not in the mapping).

**Comment by dianedetoeuf** [comment]
I've added some missing fields (and there might be others) but for now running into a new type of error about conflict with foreign key constraint: https://www.openfn.org/projects/p5x4g4/runs/rk53knjp

**Comment by lakhassane** [comment]
Hey @dianedetoeuf it seems like the column `WCSPROGRAMS_VegetationPhysiographyID` in that table is empty while it shouldn't because that's a foreign key. Below is the query that should fetch the ID for that column but it returns empty because there is no `Plain (flat)`in the database (but there is `plain`?)

```sql
select WCSPROGRAMS_VegetationPhysiographyID from WCSPROGRAMS_VegetationPhysiography where WCSPROGRAMS_VegetationPhysiographyExtCode = 'Plain (flat)'
```

**Comment by aleksa-krolls** [comment]
@dianedetoeuf Following up on @lakhassane's comment... it looks like `Plain (flat)` matches the `Name` (not ExtCode) of a Physiography record... 
<img width="517" alt="Screen Shot 2021-06-15 at 9 05 26 AM" src="https://user-images.githubusercontent.com/16758106/122001510-18951780-cdb9-11eb-82c7-eff61cbf6716.png">

So you could consider changing L58 to reference `WCSPROGRAMS_VegetationPhysiographyName` instead.
https://github.com/OpenFn/ConSoSci/blob/b29e74225c73c9716b805279064beb91c7315bac/vegetation/legacyVegetation_vDD.js#L54-L61

**Comment by dianedetoeuf** [comment]
Ok thanks, I've fixed that ! How come that it seems to work well with ExtCode in the other job (VegetationClassficationAndTreeMeasurementForm.js) ?

**Comment by dianedetoeuf** [comment]
I have the same issue with ForestType, but here changing the ExtCode to Name doesn't solve it, see https://www.openfn.org/projects/p5x4g4/runs/rbgzywg9. (We've updated all the lookup tables in the DB)

**Comment by aleksa-krolls** [comment]
@dianedetoeuf You'll need to review the DB reference data to see what columns (e.g., `Name`, `ExtCode`, `Code`) can be used to link the survey tables with these reference tables.  
For example, with `ForestType`... it doesn't look like `ExtCode` is filled in for all records... and because the Kobo value is `Open Grassland`, maybe it will be better to use the `Name` or `Code` column.  
<img width="825" alt="Screen Shot 2021-06-15 at 5 41 47 PM" src="https://user-images.githubusercontent.com/16758106/122073413-1060ca80-ce01-11eb-98a6-ba7b4b26562f.png">

**Comment by dianedetoeuf** [comment]
Thanks, Name gives me the same issue and Code gives a completely different error: Cannot read property '0' of undefined. (https://www.openfn.org/projects/p5x4g4/runs/rg5ry6bg)

**Comment by aleksa-krolls** [comment]
@dianedetoeuf again, it doesn't look like `Open Grassland` doesn't exist as a record on the ForestType table ... would you like us to add it so that you can map using `Name`? 

@lakhassane why is this TypeError being thrown here? Not an error I'd expect.

**Comment by dianedetoeuf** [comment]
hmm ok I thought we had added all the new values, will check again!

**Comment by lakhassane** [comment]
@aleksa-krolls @dianedetoeuf I don't think the job is executed on the right payload.
That one has no `surveys` array and the others (`general`, `liana_old`) aren't there neither

**Comment by aleksa-krolls** [comment]
@dianedetoeuf yes, now sure how... but it looks like you ran this legacy job against BNS data? 
<img width="738" alt="Screen Shot 2021-06-16 at 1 27 28 PM" src="https://user-images.githubusercontent.com/16758106/122203636-e82ba700-cea6-11eb-939f-edb8054d49a8.png">

Here is the message to be testing against: 
https://www.openfn.org/projects/p5x4g4/messages/mgk5v7dz

**Comment by dianedetoeuf** [comment]
Ha don't know how this happened but havn't changed anything to the trigger... And the left panel in the job writting still shows me the right plant legacy data.

**Comment by aleksa-krolls** [comment]
@dianedetoeuf I think what happened is a bug... when you clicked `Save & Run`, the job was executed against the wrong Message state (BNS message instead of Veg data). I've reported it to the product team to check out. cc @taylordowns2000 

In the meantime, please continue to let me and @lakhassane know if you encounter further issues and questions during your job testing.

**Comment by dianedetoeuf** [comment]
Hi, I'm still lost in the parenthesis for the findvalue function on all the TaxaID and observers... @lakhassane could you maybe have a look at [the job](https://www.openfn.org/projects/p5x4g4/jobs/jywrew/edit?runId=rd7m6gba) please? Thanks!

**Comment by lakhassane** [comment]
Hey @dianedetoeuf could you indicate me which line I should look at. I don't see `observers` in the current job and for `TaxaID` the parenthesis looks correct.

**Comment by dianedetoeuf** [comment]
Sure, for the observer its lines 214, 221 and 228, and for the parenthesis I have red crosses on lines 166, 174, 183, 191, 199 and 207.

**Comment by aleksa-krolls** [comment]
@lakhassane Did you see Diane's latest comment above? @dianedetoeuf feel free to coordinate with Mamadou if helpful to jump on a call to walk through your last issues. I'm still off and in the US for part of this week, but you two can certainly sync, and @ritazagoni can join if additional Kobo context is needed.

**Comment by lakhassane** [comment]
Hey @dianedetoeuf I was looking at the wrong place. So what you might be missing for the red crosses is the fact that if you want to use `await findValue({...})` you have to put it inside an `alterState(async ...)` just like the draft we made for the operation starting line `15`.
I corrected the 4 last operations starting from line `160` so you could see.

**Comment by dianedetoeuf** [comment]
Great, thank you @lakhassane !

**Comment by dianedetoeuf** [comment]
Hi, I'm running into a new issue as some TaxaID fields are null and we'd like it to transform those nulls into "unknown". Tried adding something for this transformation line 11 but not sure how to insert that in the job ?

**Comment by lakhassane** [comment]
Hey @dianedetoeuf is it better if we schedule a meeting tomorrow and work on that issue together?

**Comment by dianedetoeuf** [comment]
Hi, thank you, and yes lets discuss today on a call!

**Comment by dianedetoeuf** [comment]
Weirdly the same for native_tree_shrub doesn't seem to be working...

**Comment by lakhassane** [comment]
@dianedetoeuf can I see the output? Or we go back to zoom? I tested your code and Im seeing `unknown`

**Comment by lakhassane** [comment]
what do you have if you add this code in line 24 `console.log(state.data.surveys[0].native_tree_shrubs)`

**Comment by dianedetoeuf** [comment]
The run is here with console log added line 24 : https://www.openfn.org/projects/p5x4g4/runs/ryapajpz

**Comment by lakhassane** [comment]
It looks correct. If you look at the message you are running you have `"Native_tree_Shrub": "Dichrostachys cinerea",`. Since it's not null it won't be set to `unknown`

**Comment by dianedetoeuf** [comment]
okay but then why there is still this error : "TypeError [Error]: Cannot read property 'native_tree_shrubs' of undefined" at the end?

**Comment by lakhassane** [comment]
@dianedetoeuf we were missing this `(state)` after the closing parenthesis of the `upserMany` :/ You can re-run and check the new error (probably because you need to apply the unknown logic to the rest of the arrays.)
<img width="301" alt="Screen Shot 2021-07-07 at 15 32 41" src="https://user-images.githubusercontent.com/8411231/124788064-92dc3600-df38-11eb-8528-d858c0493ab5.png">

**Comment by dianedetoeuf** [comment]
ok not sure where I need to apply this again as the error still seems to be on the VegetationGrass that take data from the ground_species : https://www.openfn.org/projects/p5x4g4/runs/r3x3x9w6

is it saying that it doesn't recognize the "unknown"?

**Comment by lakhassane** [comment]
@dianedetoeuf from the error it seems like we are trying to insert `Null` in a column that SHOULD NOT be null (which is: `WCSPROGRAMS_VegetationGrassName`).

Now if we look closely to the mapping from line `151` to `174`about `WCSPROGRAMS_VegetationGrass` that column we talked about above is absent, so when the upsert will run the database is going to put it to null which is incorrect thus the error.

So basically in your mapping you should also map the column `WCSPROGRAMS_VegetationGrassName`.

**Comment by lakhassane** [comment]
Want us to chat about it through zoom?

**Comment by dianedetoeuf** [comment]
sure! https://us02web.zoom.us/j/86166669972?pwd=akpiVElLRGFGTllSOHMySjN4bUh5Zz09

**Comment by dianedetoeuf** [comment]
Problem with merging now : https://www.openfn.org/projects/p5x4g4/runs/rg5g5n9a

**Comment by lakhassane** [comment]
Hmm this one is a bit trickier. let me jump on the code.

**Comment by lakhassane** [comment]
@dianedetoeuf it might be more complex. The issue is that we try to insert multiple time the same row because all the objects in the `ground_species` array have the same `surveyid`. 

I tried building a unique `AnswerID` using this `${data.surveyid}${data.Ground_Spp_No}`. But Im not sure if it's the right value for the database.
cc @aleksa-krolls

**Comment by dianedetoeuf** [comment]
Green check!!! and now I'm going to open a bottle of champaign 🍾

**Comment by lakhassane** [comment]
😂 amen! Have a sip for me 🎉

**Comment by dianedetoeuf** [comment]
Hi again, the message that is being processed by the job has only 1 survey (https://www.openfn.org/projects/p5x4g4/messages/mgk5v7dz) - do you know which is the message with all the surveys?

**Comment by aleksa-krolls** [comment]
@dianedetoeuf I think this is the right Message: https://www.openfn.org/projects/p5x4g4/messages/m8qvjvbr

You can filter the inbox by the Trigger `Uganda Legacy Plant Data` to see matching Messages - see here: https://www.openfn.org/projects/p5x4g4/messages?dateRange=4320&endDate=2021-07-08T12%3A47%3A21.786Z&jobStates=&page=1&perPage=10&searchString=&startDate=2021-01-09T12%3A47%3A21.786Z&triggerId=ty5v8g

Thanks @lakhassane for all the help :)

**Comment by dianedetoeuf** [comment]
Thanks Aleksa! I have a new error about ground_species not iterable... https://www.openfn.org/projects/p5x4g4/runs/rzdpwwme

**Comment by aleksa-krolls** [comment]
@dianedetoeuf @lakhassane is looking into it... now that we're using a different Message as state (that has a lot of different survey inputs, rather than just the 1... we might need to make some other small adjustments so that the job will support all the versions of state).

**Comment by lakhassane** [comment]
Hey @dianedetoeuf I have rework the job to take into account the fact that we should insert arrays and not only one element.
What we did so far:
1. iterate through the `surveys` array and build different arrays (`general`, `ground_species`, `native_tree_shrubs`). 
2. Use those arrays in the next operations to `upsertMany`

So the job remains quite similar to what you had before.

**Comment by dianedetoeuf** [comment]
Hi, thank you! I've run the message https://www.openfn.org/projects/p5x4g4/messages/m8qvjvbr and have a new error : "The MERGE statement conflicted with the FOREIGN KEY constraint "FK_WCSPROGRAMS_Vegetation_WCSPROGRAMS_VegetationSoilDescriptionID". The conflict occurred in database "wcsprograms", table "dbo.WCSPROGRAMS_VegetationSoilDescription", column 'WCSPROGRAMS_VegetationSoilDescriptionID'."

Line 21052 of https://www.openfn.org/projects/p5x4g4/runs/rwa8zd5z

**Comment by lakhassane** [comment]
It's likely that we have duplications in the arrays and we try to insert the same row twice (cc @aleksa-krolls).

I remember in some cases we build a unique array of what we want to insert before going ahead. Should we do the same here? Using which fields to check for uniqueness?

**Comment by lakhassane** [comment]
Oh ignore the above comment actually. I just saw the constraint on the column `WCSPROGRAMS_VegetationSoilDescriptionID`. We have a foreign key constraint on that column and actually it seems like we are inserting data on that column for table `WCSPROGRAMS_Vegetation` that does not exist in the parent table (`WCSPROGRAMS_VegetationSoilDescription`) @dianedetoeuf

**Comment by lakhassane** [comment]
@dianedetoeuf for more investigation in the job we have those `findValue` operations that look for foreign key. One example is this one: 
```js
  WCSPROGRAMS_VegetationSoilDescriptionID: await findValue({
          uuid: 'WCSPROGRAMS_VegetationSoilDescriptionID',
          relation: 'WCSPROGRAMS_VegetationSoilDescription',
          where: {
            WCSPROGRAMS_VegetationSoilDescriptionName: state.handleValue(
              data['Soil description']
            ),
          },
        })(state),
```

However, in the message sometimes `Soil Description` is set to null thus the `findValue` will proly return `undefined` instead of an existing value in table `WCSPROGRAMS_VegetationSoilDescription`. Then we try to insert undefined when we shouldn't. Is there a default value to use for each column that have a foreign key constraint?

**Comment by aleksa-krolls** [comment]
@lakhassane To clarify, it looks like we can't insert a `null` value into the column `WCSPROGRAMS_VegetationSoilDescriptionID`?  If there is no value to map, then I'd assume we don't need to `findValue(...)` and can just return `undefined` as a value, no? 

@dianedetoeuf Should `WCSPROGRAMS_VegetationSoilDescriptionID` be a required column? (And if yes, is there a default value (e.g., `No Soli`) for records that don't have data for this? 
OR... can records be inserted with no `WCSPROGRAMS_VegetationSoilDescriptionID` specified?

**Comment by lakhassane** [comment]
@aleksa-krolls no we can't return `undefined` (that's the case currently) because it's a foreign key column so it should always map a value that exists in the parent table. Plus that's a` Not Null `column also

**Comment by aleksa-krolls** [comment]
Got it, thanks. So @dianedetoeuf looks like the DB requires soil description... therefore you can either create a `Not Defined` record or something like that that lives in the reference table `WCSPROGRAMS_VegetationSoilDescription` so that you have something to relate these new Vegetation records to... 
Or you can ask Steve if it's okay to remove the NOTNULL() constraint from the column `WCSPROGRAMS_VegetationSoilDescriptionID` to allow for this to be blank.

**Comment by dianedetoeuf** [comment]
Hi, on my new job for the 2nd vegetation form (most recent data), I have an error about converting data type varchar to numeric but don't know which field this refers to - how do I know that? See run here : https://www.openfn.org/projects/p5x4g4/runs/r7xnbdp6

**Comment by dianedetoeuf** [comment]
For the Legacy job, we've added "unknown" in the soil table, and I've added the code to replace nulls by unknown but I still have the issue on the FK constraint. Not sure my code to replace the null is right as the field "Soil description" as a space rather than an "_" between the 2 words but it did not seem to work with the space so I've used "Soil_description".
See line 31 of https://www.openfn.org/projects/p5x4g4/jobs/jywrew
And line 21052 of https://www.openfn.org/projects/p5x4g4/runs/ryapmmew
Thanks!

**Comment by aleksa-krolls** [comment]
@taylordowns2000 while Mamadou is off this week, can you please help Diane troubleshoot? I think adding some logs could also help out. 
- See her run failure (https://www.openfn.org/projects/p5x4g4/runs/ryapmmew ) when connecting with the `WCS Programs Database (central db)`
- See here for the list of available `WCSPROGRAMS_VegetationSoilDescription` records in the DB that we're trying to look up to: https://www.openfn.org/projects/p5x4g4/runs/r9w93dpr

Fyi, if the `Soil_description` is not provided in state, we want to map the `WCSPROGRAMS_VegetationSoilDescription` record where `WCSPROGRAMS_VegetationSoilDescriptionName = 'Unknown'`. 
https://github.com/OpenFn/ConSoSci/blob/eb4f37a97d64f105a579f90b928781de28186684/vegetation/legacyVegetation_vDD.js#L31-L34

**Comment by taylordowns2000** [comment]
Hello @dianedetoeuf ! This is great, and your instinct was spot on... I think it _is_ the space that's causing you problems.

## tl:dr;
replace
```js
if (soil.Soil_description === null) soil.Soil_description = 'unknown';
```
with
```js
if (soil['Soil description'] === null) soil['Soil description'] = 'unknown';
```

## better understanding

Check out how to access properties in javascript: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Property_accessors#syntax

Using that second style allows for much more flexibility... otherwise Node will get confused by those spaces!

Does this solve the problem?

**Comment by dianedetoeuf** [comment]
Hi, thanks Taylor! It seems to work and so I added some more nulls replacement in other columns (soil colour, moisture, erodability, seasonality). Stilll have an error about FK constraint for soil seasonality, see run here: https://www.openfn.org/projects/p5x4g4/runs/r6g7b9p4

**Comment by taylordowns2000** [comment]
Fantastic. OK, for this next one the error I'm reading is `The MERGE statement conflicted with the FOREIGN KEY constraint "FK_WCSPROGRAMS_Vegetation_WCSPROGRAMS_VegetationSoilSeasonalityID". The conflict occurred in database "wcsprograms", table "dbo.WCSPROGRAMS_VegetationSoilSeasonality", column 'WCSPROGRAMS_VegetationSoilSeasonalityID'.`

For the following upsert:
```
Executing upsert via : MERGE WCSPROGRAMS_Vegetation AS [Target]
        USING (VALUES [--REDACTED--]) AS [SOURCE] (Answer_ID, WCSPROGRAMS_VegetationName, SubmissionDate, StartTime, WCSPROGRAMS_VegetationDistrictID, UserID_CR, UserID_LM, Surveydate, SurveySite, TransectNo, PlotNumber, East, North, Latitude, Longitude, WCSPROGRAMS_VegetationAgeID, WCSPROGRAMS_VegetationClassID, WCSPROGRAMS_VegetationForestTypeID, WCSPROGRAMS_VegetationPhysiographyID, WCSPROGRAMS_VegetationTopographyID, WCSPROGRAMS_VegetationDrainageID, WCSPROGRAMS_VegetationSoilDescriptionID, WCSPROGRAMS_VegetationSoilColorID, IsEvidenceOfFire, WCSPROGRAMS_VegetationSoilMoistureID, WCSPROGRAMS_VegetationSoilErodabilityID, WCSPROGRAMS_VegetationSoilSeasonalityID, Bareground)
        ON [Target].[--VALUE--] = [Source].[--VALUE--]
        WHEN MATCHED THEN
          UPDATE SET [--REDACTED--] 
        WHEN NOT MATCHED THEN
          INSERT (Answer_ID, WCSPROGRAMS_VegetationName, SubmissionDate, StartTime, WCSPROGRAMS_VegetationDistrictID, UserID_CR, UserID_LM, Surveydate, SurveySite, TransectNo, PlotNumber, East, North, Latitude, Longitude, WCSPROGRAMS_VegetationAgeID, WCSPROGRAMS_VegetationClassID, WCSPROGRAMS_VegetationForestTypeID, WCSPROGRAMS_VegetationPhysiographyID, WCSPROGRAMS_VegetationTopographyID, WCSPROGRAMS_VegetationDrainageID, WCSPROGRAMS_VegetationSoilDescriptionID, WCSPROGRAMS_VegetationSoilColorID, IsEvidenceOfFire, WCSPROGRAMS_VegetationSoilMoistureID, WCSPROGRAMS_VegetationSoilErodabilityID, WCSPROGRAMS_VegetationSoilSeasonalityID, Bareground) VALUES [--REDACTED--];
```

Off the bat, it would be helpful to know more about _both_ the query being sent and the database you're sending it to. For the query, you could add `{ writeSql: true }` as a fourth argument for your `upsertMany` like this:
```js
return upsertMany(
    'WCSPROGRAMS_VegetationVegetationObserver',
    'Answer_ID',
    () => general,
    { writeSql: true } // <-- options go here!
  )(state);
```
And that should allow you to view all the actual SQL queries that get generated. Once you can see the values in that query, you might be able to figure out why it doesn't gel with however your MSSQL database is configured.

On the `The MERGE statement conflicted with the FOREIGN KEY constraint` side of things, do you know who set up this DB and could you loop them in? I'd love to have someone with MSSQL knowledge take a crack at diagnosing this. That said, my quick Googling makes me think that this error will arise if you're trying to add `WCSPROGRAMS_VegetationVegetationObserver` records with a `FK_WCSPROGRAMS_Vegetation_WCSPROGRAMS_VegetationSoilSeasonalityID` value that is *_not present_* as a foreign key for a record in your `WCSPROGRAMS_VegetationSoilSeasonality` table.

So... off the cuff, if you have access to the DB maybe check to see what records you've got in that `WCSPROGRAMS_VegetationSoilSeasonality` table and then check to see what IDs you're trying to set in that `FK_WCSPROGRAMS_Vegetation_WCSPROGRAMS_VegetationSoilSeasonalityID` column... if you're trying to set an ID there which doesn't have a corresponding record in the first table you'll likely get this error.

I hope this is a good start! Please keep the questions coming :-)

**Comment by dianedetoeuf** [comment]
okay thank you, not sure I understand everything but will try the write sql thing and see what I've got. Steve did set up the DB, how can we loop him in in this conversation?

**Comment by taylordowns2000** [comment]
OK, great! Re: Steve, it doesn't look like he's been active on this thread... he may have GitHub, but I'm not sure. You could always send him a direct link to the comment: https://github.com/OpenFn/ConSoSci/issues/117#issuecomment-884839573 . How do you usually handle questions related to this database?

My guess is still that it comes down to what you've got in that `WCSPROGRAMS_VegetationSoilSeasonality` table. Do you have any way of browsing what's in that table? All else failing, you could write a new job that _only_ had the following:
```js
sql(state => 'select * from dbo.WCSPROGRAMS_VegetationSoilSeasonality';)
```
and check the output!

**Comment by dianedetoeuf** [comment]
Yes Steve is on Github as wcs-sgallo :)

In theory all the values from the legacy data should be in the lookup tables but will check again... thanks!

**Comment by taylordowns2000** [comment]
Ok brilliant. I'll at mention him here! Do you know what value you're trying to put in `WCSPROGRAMS_VegetationSoilSeasonalityID` when the job fails?

@wcs-sgallo , in this job @dianedetoeuf is having trouble with a foreign key constraint: `The MERGE statement conflicted with the FOREIGN KEY constraint "FK_WCSPROGRAMS_Vegetation_WCSPROGRAMS_VegetationSoilSeasonalityID". The conflict occurred in database "wcsprograms", table "dbo.WCSPROGRAMS_VegetationSoilSeasonality", column 'WCSPROGRAMS_VegetationSoilSeasonalityID'`

My assessment is that _probably_ means she's trying to add a record that points to a non-existent `WCSPROGRAMS_VegetationSoilSeasonality` record, but I don't know much about how this DB is set up (and I'm no MSSQL-master myself.) Do you have any thoughts?

**Comment by dianedetoeuf** [comment]
Yeah so I thought the problem was coming from the nulls that I replaced by the value "unknown", but Steve already added unknown to all those tables. And we added the other values also. No idea which value is causing the issue, should I be able to see that in the run?

**Comment by dianedetoeuf** [comment]
It looks like the last value the job is trying to import for seasonality is "Wet" but we do have this value in the table.

**Comment by taylordowns2000** [comment]
Understood. OK, so I've sent a new message to the inbox that's a little easier to work with... about 80 lines of JSON instead of 330,000!

I've also added a console.log so you can see the data you're about to upsert. Check out the logs here: https://www.openfn.org/projects/p5x4g4/runs/rwa8xaww

Maybe it would be easiest to work with individual submissions like this (or several, instead of thousands) until you're happy with how they're mapping (taking a peek at the output each time) and _then_ try to process that big 330,000 line payload. This one happens to be upset about `WCSPROGRAMS_VegetationForestTypeID`, but it also looks like that might not be getting handled as expected as I see the output as undefined in `line 44` of that run.

**Comment by aleksa-krolls** [comment]
@taylordowns2000 Wondering if we can go ahead and split up the 1 massive Message into individual Messages? Because I'm sure this is not the only survey with issues. 
- This is the master Message with all the legacy survey csv data: https://www.openfn.org/projects/p5x4g4/messages/m8qvjvbr
- This is the trigger all the Messages need to match: https://www.openfn.org/projects/p5x4g4/triggers/ty5v8g

@dianedetoeuf This will allow you to bulk process all the Messages and see which surveys have synced and which ones still have issues to more effectively troubleshoot. 

Also, doesn't look like `Wetland floats` is an available record in your reference table `WCSPROGRAMS_VegetationForestType`, which is why it's returning as `undefined` on L44 of the run log: https://www.openfn.org/projects/p5x4g4/runs/rwa8xaww

**Comment by dianedetoeuf** [comment]
Ok I can try with individual messages then. Is it possible to archive the message that contains all the data then so that the job doesn't try it everytime? Thanks!

**Comment by aleksa-krolls** [comment]
Yes @dianedetoeuf ... I changed the tags in the "master" Message that contains all the data. You can now filter Inbox using this trigger. 
![Screen Shot 2021-07-23 at 12.36.23 PM.png](https://images.zenhubusercontent.com/5d691cd53295e300013f1cbd/c32d19b9-4491-48ca-b703-6ddc482352a8)
And as Taylor creates the new Messages, he will include the tag below so you can keep filtering by this only to test or bulk re-process. 
![Screen Shot 2021-07-23 at 12.37.54 PM.png](https://images.zenhubusercontent.com/5d691cd53295e300013f1cbd/b99830f1-13cd-4cd0-965e-24e75ad4e63e)

**Comment by taylordowns2000** [comment]
Hi @dianedetoeuf and @aleksa-krolls , they're all in there! Use this setup to select them:

https://www.openfn.org/projects/p5x4g4/messages?dateRange=24&endDate=2021-07-23T13%3A56%3A33.963Z&jobStates=&page=1&perPage=10&searchString=&startDate=2021-07-22T13%3A56%3A33.963Z&triggerId=ty5v8g

and then click re-process and you should see a preview interface like this:

![image.png](https://images.zenhubusercontent.com/566ac5e0638e718c995f5356/dbb2f357-7f26-4fba-a428-6255c1ae2c77)
